python/tvm/meta_schedule/cost_model/mlp_model.py

```diff
@@ -0,0 +1,743 @@
+        """LambdaLoss: Metric-Driven Loss for Learning-to-Rank
+
+        Parameters
+        ----------
+        preds : Tensor
+            The predicted runtime for each candidate.
+        labels : Tensor
+            The measured runtime for each candidate.
+        k : int
+            Loss for top k.
+            Default is None, which means computing all scores.
+        eps : float
+            The minimum value to the denominator and argument of log if they reach 0.
+        sigma : float
+            The scaling factor to the input of the sigmoid function.
+        """
+        device = preds.device
+        y_pred, y_true = preds[None, :], labels[None, :]
+        y_pred_sorted, indices_pred = y_pred.sort(descending=True, dim=-1)
+        y_true_sorted, _ = y_true.sort(descending=True, dim=-1)
+        true_sorted_by_preds = torch.gather(y_true, dim=1, index=indices_pred)
+        true_diffs = true_sorted_by_preds[:, :, None] - true_sorted_by_preds[:, None, :]
+        padded_pairs_mask = torch.isfinite(true_diffs) & (true_diffs > 0)
+        ndcg_at_k_mask = torch.zeros(
+            (y_pred.shape[1], y_pred.shape[1]), dtype=torch.bool, device=device
+        )
+        ndcg_at_k_mask[:k, :k] = 1
+        true_sorted_by_preds.clamp_(min=0.0)
+        y_true_sorted.clamp_(min=0.0)
+        pos_idxs = torch.arange(1, y_pred.shape[1] + 1).to(device)
+        D = torch.log2(1.0 + pos_idxs.float())[None, :]
+        maxDCGs = torch.sum(((torch.pow(2, y_true_sorted) - 1) / D)[:, :k], dim=-1).clamp(min=eps)
+        G = (torch.pow(2, true_sorted_by_preds) - 1) / maxDCGs[:, None]
+        weights = torch.abs(
+            torch.pow(D[:, :, None], -1.0) - torch.pow(D[:, None, :], -1.0)
+        ) * torch.abs(G[:, :, None] - G[:, None, :])
+        scores_diffs = (y_pred_sorted[:, :, None] - y_pred_sorted[:, None, :]).clamp(
+            min=-1e8, max=1e8
+        )
+        scores_diffs[torch.isnan(scores_diffs)] = 0.0
+        weighted_probs = (torch.sigmoid(sigma * scores_diffs).clamp(min=eps) ** weights).clamp(
+            min=eps
+        )
+        losses = torch.log2(weighted_probs)
+        masked_losses = losses[padded_pairs_mask & ndcg_at_k_mask]
+        loss = -torch.sum(masked_losses)
+        return loss
+
+    def topk_score(
+        self,
+        pred_results: torch.Tensor,
+        gt_results: torch.Tensor,
+        k: int,
+    ) -> float:
+        """
+        Evaluate the top-k score
+
+        Parameters
+        ----------
+        pred_results: Tensor
+            The raw prediction
+        gt_results: Tensor
+            The measured results
+        k : int
+            The k in top k score
+
+        Returns
+        -------
+        score : float
+            The top-k score
+        """
+        topk_indices = torch.topk(pred_results, k, largest=False).indices
+        score = gt_results.min() / gt_results[topk_indices].min()
+        return score.item()
+
+
+@derived_object
+class MLPModel(PyCostModel):
```

Comment:
Let's do some slight refactoring moving code around to make sure it's easy enough to use in both pre-training and MetaSchedule integration.

In this file, we will have:

```python
class SegmentDataLoader: ...

class SegmentSumMLP(torch.nn.Module):
  def __init__(...): ...
  def forward(...): ...
  def lambda_rank_loss(...): ...
  def topk_score(...): ...

class SegmentSumMLPTrainer:
  # this handles data loading and training of `SegmentSumMLP`

class MLPModel(PyCostModel):
  model: SegmentSumMLP

  def __init__(...): ...
  def load(...): ...
  def save(...): ...
  def update(...): ... # an empty method if the model is frozen
  def predict(...):  ... # calls `self.model` to make predictions
  <del>def _train(...): ... </del> # move to the Trainer; no need to implement this method
  <del>def _validate(...): ... </del> # move to the Trainer; no need to implement this method
```

The reason that I suggested such refactoring is because I would love to make sure the two logics are decoupled:
- Model pretraining
- Integration with MetaSchedule

With the `Trainer`, the logic for model pretraining doesn't need `MLPModel` anymore, and `MLPModel` can focus on the integration part without having to worry about the model training.

Reply:
Updated a draft for this.. Perhaps can discuss a bit?

Reply:
Addressed :-)